### PR TITLE
[loader V2] Full diff for easier reviewing

### PR DESCRIPTION
**Pull request by georgemitenkov**
## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-vm/runtime/src/storage/environment.rs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        compiled_module: Arc<CompiledModule>,
+        module_size: usize,
+        module_hash: &[u8; 32],
+    ) -> VMResult<PartiallyVerifiedModule> {
+        if !VERIFIED_MODULES_V2.contains(module_hash) {
+            // For regular execution, we cache already verified modules. Note
+            // that this even caches verification for the published modules.
+            // This should be ok because as long as the hash is the same, the
+            // deployed bytecode and any dependencies are the same, and so the
+            // cached verification result can be used.
```
Comment (zekun000) [inline third_party/move/move-vm/runtime/src/storage/environment.rs:173]: is this true? module hash doesn't include information about dependencies iiuc. also the verification result could change with bytecode verifier version/config update

File: third_party/move/move-vm/runtime/src/storage/environment.rs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        sha3_256.update(bytes);
+        let module_hash: [u8; 32] = sha3_256.finalize().into();
+
+        Ok((compiled_module, bytes.len(), module_hash))
+    }
+
+    /// Returns ann error is module's address and name do not match the expected values.
+    /// In general, we enforce this is the case at module publish time.
+    #[inline]
+    pub fn paranoid_check_module_address_and_name(
```
Comment (zekun000) [inline third_party/move/move-vm/runtime/src/storage/environment.rs:237]: is this a new check added?

File: aptos-move/block-executor/src/modules_and_scripts.rs [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+                }
+
+                // Otherwise, we need to go to the multi-version data structure to get it, and
+                // record under captured reads.
+                let read = state
+                    .versioned_map
+                    .code_storage()
+                    .module_storage()
+                    .get_or_else(key, self.txn_idx, || {
+                        self.get_base_module_storage_entry(key)
```
Comment (zekun000) [inline aptos-move/block-executor/src/modules_and_scripts.rs:259]: this may destroy performance when every txn is trying to read 0x1::framework and all go into the base version and do deserialization/verification. just a note

File: aptos-move/block-executor/src/modules_and_scripts.rs [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+    /// Similar to [LatestView::read_module_storage], but in case the module does not exist,
+    /// returns a linker VM error.
+    fn get_existing_module_storage_entry_with_version(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> VMResult<(ModuleVersion, Arc<ModuleStorageEntry>)> {
+        self.read_module_storage(address, module_name)?
+            .into_versioned()
+            .ok_or_else(|| module_linker_error!(address, module_name))
```
Comment (zekun000) [inline aptos-move/block-executor/src/modules_and_scripts.rs:302]: linker error feels weird, but I guess that's consistent with existing behavior?

File: third_party/move/move-vm/runtime/src/storage/loader.rs [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+    pub(crate) fn ty_builder(&self) -> &TypeBuilder {
+        &self.vm_config().ty_builder
+    }
+
+    pub(crate) fn ty_cache(&self) -> &RwLock<TypeCache> {
+        &self.ty_cache
+    }
+
+    pub(crate) fn struct_name_index_map(&self) -> &StructNameIndexMap {
```
Comment (zekun000) [inline third_party/move/move-vm/runtime/src/storage/loader.rs:74]: how does this work if we need to invalidate? this is still an append only data structure right? 

are we only making published modules visible during rolling commit? unlike resource writes becomes visible immediately after execution

File: aptos-move/block-executor/src/modules_and_scripts.rs (hunk 4 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/modules_and_scripts.rs:302]: +1, I think we need Result<Option<V>> interfaces everywhere, and caller decides to map it to linker error or not. This way we also avoid multiple calls to executor.

File: third_party/move/move-vm/runtime/src/storage/loader.rs (hunk 5 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/storage/loader.rs:74]: Yes, published modules are only visible after rolling commit. For init module, we start a new VM so that we don't cache what needs invalidation later.

Btw, I realised new VM makes V2 incompatible with V1 because finishing session before init module messes up gas charges for cached resources. I will move struct name index map (runtime env) to block executor and VM should not know about it. This should be not many changes.

File: third_party/move/move-vm/runtime/src/storage/environment.rs (hunk 2 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/storage/environment.rs:237]: Old check, encapsulated as a function to share between V1 and V2.

File: third_party/move/move-vm/runtime/src/storage/environment.rs (hunk 1 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/storage/environment.rs:173]: Hash includes all binary, so even dependency handles? Here verification is local, so we do not know if this module is ok wrt dependencies, but locally same hash should guarantee the equivalence?

This does not work with changes in verifier, but current code also does not work with that iirc @runtian-zhou? Ideally, we avoid this cache by having a single cache outside of the block which captures verification per epoch.

Added it to compare to V1 loader because if it is not there, V1 is faster :(

Comment (zekun000) [reply third_party/move/move-vm/runtime/src/storage/environment.rs:173]: hash includes dependency identifier but not dependency content right? i assume compatibility rules enforce once it's published, dependency upgrade can't break it but want to double check

File: aptos-move/block-executor/src/modules_and_scripts.rs (hunk 3 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/modules_and_scripts.rs:259]: Yep, we should do it under lock, I think @gelash suggested that as well. We can just add a flag there, but what do we do with all other threads? Wait? Ideally we just let them do some other useful work while they wait.

File: third_party/move/move-vm/runtime/src/storage/loader.rs (hunk 5 above)
Comment (zekun000) [reply third_party/move/move-vm/runtime/src/storage/loader.rs:74]: wondering how do we support module eviction with this struct later if we have a long running base cache for modules

Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/storage/loader.rs:74]: For every evicted module, once can go over it struct types and evict the entries in name cache, but that would need to keep some reference counting I guess because there can be other modules using it.

File: third_party/move/tools/move-unit-test/src/test_runner.rs [hunk 6]
```diff
@@ -75,23 +77,43 @@
...
+        module_storage.add_module_bytes(
+            module.self_addr(),
+            module.self_name(),
+            module_bytes.into(),
+        );
     }
 
-    Ok(storage)
+    if let Some(genesis_state) = genesis_state {
+        // In V2 loader implementation, we differentiate between modules and resources, so
```
Comment (runtian-zhou) [inline third_party/move/tools/move-unit-test/src/test_runner.rs:102]: Is the problem that the move unit test setup stores modules and resources differently? Seems to me that they should be all viewed as bytes so I don't quite get why we need different logic here?

File: third_party/move/tools/move-bytecode-utils/src/compiled_module_viewer.rs [hunk 7]
```diff
@@ -4,11 +4,24 @@
...
+use std::{borrow::Borrow, sync::Arc};
 
 pub trait CompiledModuleView {
     type Item: Borrow<CompiledModule>;
 
     // TODO: Consider using address and module name instead of module id.
     fn view_compiled_module(&self, id: &ModuleId) -> anyhow::Result<Option<Self::Item>>;
 }
+
+impl<M: ModuleStorage> CompiledModuleView for M {
```
Comment (runtian-zhou) [inline third_party/move/tools/move-bytecode-utils/src/compiled_module_viewer.rs:17]: Is this change needed?

File: third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs [hunk 8]
```diff
@@ -150,31 +167,81 @@
...
         };
 
-        adapter
-            .perform_session_action(None, |session, gas_status| {
-                for module in either_or_no_modules(pre_compiled_deps_v1, pre_compiled_deps_v2)
-                    .into_iter()
-                    .map(|tmod| &tmod.named_module.module)
-                {
+        if vm_config.use_loader_v2 {
+            let addresses = either_or_no_modules(pre_compiled_deps_v1, pre_compiled_deps_v2)
```
Comment (runtian-zhou) [inline third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs:176]: Was wondering why we need this logic here? IIUC we try to put the modules into the local store but was wondering can we just leave them as unresolved form and load them as we go?

**Review by runtian-zhou — COMMENTED** [review]
In general I felt a bit uneasy when seeing having to add `IntoUnsyncModuleStorage, LocalModuleBytesStorage` together, particularly how the new interface is changing how we set up our test enviroment. Though adding new apis are great, but given how much change we already need just to fix our own tests, I would imagine the new loader infrastructure would bring much bigger impact if downstream want to apply the code. IMO I really hope to see the changes in our test env to be as minimal and mechanical as possible, which would indicate an easier downstream adoption process.

File: third_party/move/move-vm/transactional-tests/tests/module_publishing/republish_module_cyclic_dependencies_friend.mvir
Comment (runtian-zhou) [inline third_party/move/move-vm/transactional-tests/tests/module_publishing/republish_module_cyclic_dependencies_friend.mvir:1]: Will v1 and v2 compiler behave differently here?

File: third_party/move/move-vm/runtime/src/storage/publishing.rs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+    vm_status::StatusCode,
+};
+use std::{
+    collections::{btree_map, BTreeMap},
+    sync::Arc,
+};
+
+/// An implementation of [ModuleBytesStorage] that stores temporary changes. If used by
+/// [ModuleStorage], the most recent version of a module will be fetched.
+struct TemporaryModuleBytesStorage<'m, M> {
```
Comment (runtian-zhou) [inline third_party/move/move-vm/runtime/src/storage/publishing.rs:29]: Call this `ModulePublishingStorage` or sth to indicate that it is tightly related with module publishign logic?

File: third_party/move/move-vm/runtime/src/storage/module_storage.rs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    fn fetch_verified_module(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> VMResult<Arc<Module>>;
+}
+
+/// Storage that contains serialized modules. Clients can implement this trait
+/// for their own backends, so that [ModuleStorage] can be built on top of it.
+pub trait ModuleBytesStorage {
```
Comment (runtian-zhou) [inline third_party/move/move-vm/runtime/src/storage/module_storage.rs:70]: Seems that this trait is the same as the one defined in move_core_types?

File: third_party/move/tools/move-bytecode-utils/src/compiled_module_viewer.rs (hunk 7 above)
Comment (georgemitenkov) [reply third_party/move/tools/move-bytecode-utils/src/compiled_module_viewer.rs:17]: No, I think we can reuse InMemoryStorage here as well.

File: third_party/move/move-vm/transactional-tests/tests/module_publishing/republish_module_cyclic_dependencies_friend.mvir
Comment (georgemitenkov) [reply third_party/move/move-vm/transactional-tests/tests/module_publishing/republish_module_cyclic_dependencies_friend.mvir:1]: It's not about compiler (need to check the test case) but loader: we will allow cyclic friends, we will only check if they exist

File: third_party/move/tools/move-bytecode-utils/src/compiled_module_viewer.rs (hunk 7 above)
Comment (runtian-zhou) [reply third_party/move/tools/move-bytecode-utils/src/compiled_module_viewer.rs:17]: SG. Let's remove the unnecessary change just to make things a bit easier :D

File: third_party/move/move-vm/transactional-tests/tests/module_publishing/republish_module_cyclic_dependencies_friend.mvir
Comment (runtian-zhou) [reply third_party/move/move-vm/transactional-tests/tests/module_publishing/republish_module_cyclic_dependencies_friend.mvir:1]: Sry I mean loader. Cyclic friend is a fun issue. Do you think it would be easy to replicate the existing behavior in V2 loader?

File: third_party/move/move-vm/runtime/src/storage/module_storage.rs (hunk 10 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/storage/module_storage.rs:70]: So overall, I think we can reduce the number of changes by removing LocalModuleBytesStorage - this can be replaced by InMemoryStorage and make things nicer. Bytes trait I still need, as something that can redirect the call to StateView. Essentially, to set up any module storage you need 1) trait for module storage and 2) trait for bytes from which you can fetch into code cache (module storage) and where you store published code, essentially a StateView.

File: third_party/move/move-vm/transactional-tests/tests/module_publishing/republish_module_cyclic_dependencies_friend.mvir
Comment (georgemitenkov) [reply third_party/move/move-vm/transactional-tests/tests/module_publishing/republish_module_cyclic_dependencies_friend.mvir:1]: But we don't want to replicate the behaviour? I ll send you a doc on slack :)

File: third_party/move/tools/move-bytecode-utils/src/compiled_module_viewer.rs (hunk 7 above)
Comment (georgemitenkov) [reply third_party/move/tools/move-bytecode-utils/src/compiled_module_viewer.rs:17]: +1! For all Move tests, I will do a pass to simplify them as promised to @gelash already 😄 just wanted to put up the core logic out so that tests pass and things work ))

Comment (georgemitenkov) [reply third_party/move/tools/move-bytecode-utils/src/compiled_module_viewer.rs:17]: Will post a PR with som cleanups👍

File: aptos-move/block-executor/src/modules_and_scripts.rs (hunk 3 above)
Comment (gelash) [reply aptos-move/block-executor/src/modules_and_scripts.rs:259]: lock's implementation should put the thread to sleep and the resources can be used elsewhere in the big system, so it shouldn't be too bad.

**Comment by georgemitenkov** [comment]
@zekun000 @runtian-zhou, let's use https://github.com/aptos-labs/aptos-core/pull/14184 for all comments
